src/game/logic/system/gamelogic.cpp

```diff
@@ -62,40 +62,117 @@ bool GameLogic::Is_Intro_Movie_Playing()
     return m_startNewGame && g_theDisplay->Is_Movie_Playing();
 }
 
-int GameLogic::Rebalance_Parent_Sleepy_Update(int update)
+int GameLogic::Rebalance_Parent_Sleepy_Update(int index)
 {
-    captainslog_dbgassert(update >= 0 && (unsigned)update < m_sleepingUpdateModules.size(), "bad sleepy idx");
+    captainslog_dbgassert(index >= 0 && (unsigned)index < m_sleepingUpdateModules.size(), "bad sleepy idx");
 
-    for (int i = ((update + 1) >> 1) - 1; i >= 0; i = ((i + 1) >> 1) - 1) {
-        if (!Compare_Update_Modules(m_sleepingUpdateModules[i], m_sleepingUpdateModules[update])) {
+    for (int i = ((index + 1) >> 1) - 1; i >= 0; i = ((i + 1) >> 1) - 1) {
+
+        UpdateModule *a = m_sleepingUpdateModules[i];
+        UpdateModule *b = m_sleepingUpdateModules[index];
+
+        if (!UpdateModule::Compare_Update_Modules(a, b)) {
             break;
         }
 
-        UpdateModule *a = m_sleepingUpdateModules[i];
-        UpdateModule *b = m_sleepingUpdateModules[update];
-        m_sleepingUpdateModules[update] = a;
+        m_sleepingUpdateModules[index] = a;
         m_sleepingUpdateModules[i] = b;
-        a->Set_Index_In_Logic(update);
+        a->Set_Index_In_Logic(index);
         b->Set_Index_In_Logic(i);
-        update = i;
+        index = i;
     }
 
-    return update;
+    return index;
 }
 
-void GameLogic::Push_Sleepy_Update(UpdateModule *update)
+int GameLogic::Rebalance_Child_Sleepy_Update(int index)
 {
-    captainslog_dbgassert(update, "You may not pass null for sleepy update info");
-    m_sleepingUpdateModules.push_back(update);
-    update->Set_Index_In_Logic(m_sleepingUpdateModules.size() - 1);
+    captainslog_dbgassert(index >= 0 && (unsigned)index < m_sleepingUpdateModules.size(), "bad sleepy idx");
+
+    int next_index = 2 * index + 1;
+    UpdateModule **curr = &m_sleepingUpdateModules[index];
+    UpdateModule **next = &m_sleepingUpdateModules[next_index];
+    UpdateModule **end = &m_sleepingUpdateModules[0] + m_sleepingUpdateModules.size();

```

Comment:
m_sleepingUpdateModules[m_sleepingUpdateModules.size()]?

Reply:
That won't work, because that would access element past end and be undefined behaviour.

Reply:
I can check if I can utilize iterators maybe. `m_sleepingUpdateModules[0]` does not look very safe. If `m_sleepingUpdateModules` was empty, this would crash.

Reply:
See fixup commit added to this PR with iterator approach commit. Maybe that is better to look at.